### PR TITLE
feat: cairo 1

### PR DESCRIPTION
**Pull request by tserg**

### CODE REVIEW

**Review by bllu404 — APPROVED** [review]
LGTM!

File: README.md [hunk 1]
```diff
@@ -2,115 +2,119 @@
 
 ![tests](https://github.com/lindy-labs/cairo-wadray/actions/workflows/tests.yml/badge.svg)
 
-This library provides a felt-based implementation of arithmetic functions for two types of fixed-point decimal numbers, `wad` (18 decimals of precision) and `ray` (27 decimals of decimal numbers), written in Cairo for [StarkNet](https://www.cairo-lang.org/docs/).
-
-The library has been extensively tested with Hypothesis. For more details on running the tests, refer to this [section](#run-tests).
-
-The design of this library was originally inspired by Influenceth's [64x61 fixed-point math library](https://github.com/influenceth/cairo-math-64x61).
+This library provides a felt-based implementation of arithmetic functions for two types of fixed-point decimal numbers, `Wad` (18 decimals of precision) and `Ray` (27 decimals of decimal numbers), written in Cairo for [StarkNet](https://docs.cairo-lang.org/index.html).
```
Comment (milancermak) [inline README.md:5]: StarkNet -> Starknet

File: README.md [hunk 2]
```diff
@@ -2,115 +2,119 @@
...
+where both `a` and `b` are of the same type.
+
+There is also a set of functions for operations involving a `Wad` and a `Ray`:
+- `wmul_wr`/`wmul_rw`: Multiply a `Wad` value with a `Ray` value, and divide the product by one `Wad` scale to return a `Ray`
+- `wdiv_rw`: Scale up a `Ray` value by one `Wad` scale and divide the scaled value by a `Wad` value to return a `Ray`
+- `rmul_wr`/`rmul_rw`: Multiply a `Wad` value with a `Ray` value, and divide the product by one `Ray` scale to return a `Wad`
+- `rdiv_wr`: Scale up a `Wad` value by one `Ray` scale and divide the scaled value by a `Ray` to return a `Wad`
+- `rdiv_ww`: Scale up a `Wad` value by one `Ray` scale and divide the scaled value by a `Wad` to return a `Ray`
+
+For multiplication, the prefixes `w` and `r` denote whether the product is divided (i.e. scaled down) by a `Wad` or `Ray` respectively. For division, the prefixes `w` and `r` denote whether the first operand is multiplied (i.e. scaled up) by a `Wad` or `Ray` before the division respectively.
```
Comment (milancermak) [inline README.md:42]: Can you add a note somewhere around here about the loss of precision with some of these operations? Just to be really obvious about it.

File: README.md [hunk 3]
```diff
@@ -2,115 +2,119 @@
...
+Comparison can be performed as follows:
+- `a == b`
+- `a != b`
+- `a > b`
+- `a >= b`
+- `a < b`
+- `a <= b`
+where both `a` and `b` are of the same type.
 
 These functions operate on any `ufelt` values in the range [0, 2<sup>125</sup>], including `wad` and `ray`.
```
Comment (milancermak) [inline README.md:72]: Obsolete now.

File: README.md [hunk 4]
```diff
@@ -2,115 +2,119 @@
...
 ### Conversion
 
-- `to_uint(n)`: Convert a felt value to `Uint256`
-- `from_uint(n)`: Convert a `Uint256` value to felt with a bounds check that the result is within the range [-2<sup>125</sup>, 2<sup>125</sup>]
-- `to_wad(n)`: Multiply `n` by 10<sup>18</sup> with a bounds check that the result is within the range [-2<sup>125</sup>, 2<sup>125</sup>]
-- `wad_to_ray(n)`: Multiply `n` by 10<sup>9</sup> with a bounds check that the result is within the range [-2<sup>125</sup>, 2<sup>125</sup>]
-- `ray_to_wad(n)`: Divide `n` by 10<sup>9</sup>
-
-### Bounds Check
+The following conversions are supported by implementing the `Into` trait:
```
Comment (milancermak) [inline README.md:78]: "by implementing" sounds a little like the users still need to implement it. Maybe just saying "via the Into trait" or sth similar? Same below, with TryInto.

File: Scarb.toml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "cairo_wadray"
+version = "0.1.0"
+
+[dependencies]
+starknet = "2.3.1"
+snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry.git", tag = "v0.11.0" }
```
Comment (milancermak) [inline Scarb.toml:7]: Foundry should be a [`[dev-dependencies]`](https://docs.swmansion.com/scarb/docs/reference/specifying-dependencies.html#development-dependencies)

File: README.md [hunk 6]
```diff
@@ -2,115 +2,119 @@
...
-- `assert_result_valid(n)`: Raises an error if `n` is not in the range [-2<sup>125</sup>, 2<sup>125</sup>]
-- `assert_result_valid_unsigned(n)`: Raises an error if `n` is not in the range [0, 2<sup>125</sup>]
+The following conversions are supported by implementing the `TryInto` trait:
+- `Wad` -> `Option::<Ray>`: Multiply the `Wad` value by 10<sup>9</sup> and return `Option::Some<Ray>` if there is no overflow or `Option::None` otherwise.
+- `u256` -> `Option::<Wad>`: Returns `Option::Some<Wad>` if the value is within bounds of `u128` or `Option::None` otherwise.
 
 ## Usage
 
-To use this library, include a copy of `wad_ray.cairo` in your project, and import the library into your Cairo contracts.
+To use this library, include a copy of `wadray.cairo` in your project, and import the library into your Cairo contracts.
```
Comment (milancermak) [inline README.md:89]: The lib should be a Scarb package. Installation will then be just adding one line into `Scarb.toml` [dependencies] (and eventually, once Scarb has a package registry, even just `scarb add cairo-wadray`).

File: Scarb.toml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "cairo_wadray"
```
Comment (milancermak) [inline Scarb.toml:2]: Maybe just `wadray`? WDYT? IMO it's clear it's in Cairo.

**Review by milancermak — COMMENTED** [review]
IDK if you want to do it in this PR or a follow up one, but the library should be made into a Scarb package so it's easily installable. I think the bare minimum to do this is just adding a `[lib]` into Scarb.toml. But to be a bit more complete and helpful, it's worthwhile to add more of the [keys that can appear in the `[package]` section](https://docs.swmansion.com/scarb/docs/reference/manifest.html#package). For reference, here's [what I did](https://github.com/milancermak/xoroshiro-cairo/blob/master/Scarb.toml) in my xoroshiro lib a while ago.

I'm assuming support for Signed will also come eventually? 😇

File: README.md (hunk 4 above)
Comment (tserg) [reply README.md:78]: I will change the phrasing across the README as well for clarity.

**Comment by tserg** [comment]
> I'm assuming support for Signed will also come eventually? 

Yes, I will add it in a separate PR. Should I move the tests into a new `tests/` directory here, or wait for the signed wadray PR?

File: README.md (hunk 6 above)
Comment (tserg) [reply README.md]: Done! I will verify that the package installation works once this is merged.

**Comment by tserg** [comment]
> re: tests dir - idk, whatever works.

Okay, I will leave it to the signed PR to clean it up.

**Comment by tserg** [comment]
Going ahead to merge so I can check on the installation as Scarb package, and to unblock Signed PR.

**Comment by tserg** [comment]
I have verified that installing this library as a Scarb dependency works locally.

**Comment by milancermak** [comment]
Great stuff

**Pull request merged**